src/WebJobs.Script.WebHost/Diagnostics/EtwEventGenerator.cs

```diff
@@ -19,6 +19,11 @@ public EtwEventGenerator()
 
         public void LogFunctionTraceEvent(LogLevel level, string subscriptionId, string appName, string functionName, string eventName, string source, string details, string summary, string exceptionType, string exceptionMessage, string functionInvocationId, string hostInstanceId, string activityId, string runtimeSiteName, string slotName, DateTime eventTimeStamp)
         {
+            if (!FunctionsSystemLogsEventSource.Instance.IsEnabled())
```

Comment:
There's also another method below where we're doing a date format inline. If our goal is to avoid all unnecessary formatting when disabled, I think we'd handle that upstream on creation of the generator instance [here](https://github.com/Azure/azure-functions-host/blob/ebe1aafa9bf01ed56202b21a1a47a1897d2cebf9/src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs#L117). If etw source is disabled, the final clause would just create and register a NullEventGenerator which noops.

As far as impact of these changes, the only time we'd be creating an ETW generator with the source disabled would be in local test scenarios. In production, the ETW generator will only be created when deployed to windows, and the source will be enabled there.

So while I don't think this will have production impact, I'm not opposed to a change like this.

Reply:
I talked with @brettsam a bit this morning as well to catch up - if we're all onboard to `"{dateVariable:O}"` inline for better perf and fewer allocations, I can take a pass at that as a single PR and eliminate those items...but it won't help this specific case any since it's ETW.

Alternatively to both that and this change, I could move the date string down into the `FunctionsSystemLogsEventSource` calls within their existing IsEnabled checks (changing their signature from `string` to `DateTime`), though the code would be in a few spots - it'd be consolidated and like you'd likely expect in a disabled logger - thoughts on that as well?

Looking at this further, I see in many cases the recommended path is to pass a `long` for the `DateTime` (e.g. `.ToFileTime()`) into the `EventData` - that's a riskier change of course but any idea if that's been explore before and we could eliminate this allocation path completely? (only the `summary` would remain from `SystemLogger`)

Reply:
Sorry, just rando watching this PR because of educational interest. Here are a few thoughts. 

> Alternatively to both that and this change, I could move the date string down into the FunctionsSystemLogsEventSource calls within their existing IsEnabled checks

That seems to be the better alternative. Looking from the perspective of the event generator and the event source interaction, it seems the `IsEnabled` check that is supposed to be done inside the source is now leaking up one abstraction level higher. Changing the method signature on the source to datetime and make the formatting the responsibility of the source itself seems to be a reasonable choice, given that you then also free the callers from having to make the decision of how to format (unless that was the intent of the previous design)